### PR TITLE
Add `prefix_de` to SnapshotMap

### PR DESCRIPTION
**Pull request by maurolacy**
#461 follow-up. Adds `prefix_de` and friends to `SnapshotMap` (Addresses second part of https://github.com/CosmWasm/cw-plus/issues/461#issuecomment-961153138).

### CODE REVIEW

File: packages/storage-plus/src/snapshot/map.rs [hunk 1]
```diff
@@ -191,6 +192,28 @@ where
...
+        max: Option<PrefixBound<'a, K::Prefix>>,
+        order: cosmwasm_std::Order,
+    ) -> Box<dyn Iterator<Item = StdResult<(K::Output, T)>> + 'c>
+    where
+        T: 'c,
+        'a: 'c,
+        K: 'c,
+        K::Output: 'static,
+    {
+        let mapped = namespaced_prefix_range(store, self.primary.namespace(), min, max, order)
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot/map.rs:212]: This looks same as others

File: packages/storage-plus/src/snapshot/map.rs [hunk 2]
```diff
@@ -205,6 +228,28 @@ where
...
+        K::Output: 'static,
+    {
+        self.no_prefix_de().keys_de(store, min, max, order)
+    }
+
+    pub fn prefix_de(&self, p: K::Prefix) -> Prefix<K::Suffix, T> {
+        Prefix::new(self.primary.namespace(), &p.prefix())
+    }
+
+    pub fn sub_prefix_de(&self, p: K::SubPrefix) -> Prefix<K::SuperSuffix, T> {
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot/map.rs:249]: It is still a bit funny for me that the types here (connection between `p: K::SubPrefix` and `Prefix<K::SuperSuffix` seems to only be when writing this code and not enforced in the compile, but it is fine. Here we hide this chance to error from any user, and we are quite careful with this library code.

Mainly a thought for a future type-system

File: packages/storage-plus/src/snapshot/map.rs [hunk 3]
```diff
@@ -215,8 +260,8 @@ mod tests {
     use super::*;
     use cosmwasm_std::testing::MockStorage;
 
-    type TestMap = SnapshotMap<'static, &'static [u8], u64>;
-    type TestMapCompositeKey = SnapshotMap<'static, (&'static [u8], &'static [u8]), u64>;
+    type TestMap = SnapshotMap<'static, &'static str, u64>;
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot/map.rs:263]: :+1:

File: packages/storage-plus/src/snapshot/map.rs [hunk 4]
```diff
@@ -492,9 +511,27 @@ mod tests {
...
-                ((b"C".to_vec(), b"A".to_vec()), 22)
+                (("B".into(), "B".into()), 13),
+                (("C".into(), "A".into()), 22)
             ]
         );
     }
+
+    #[test]
+    #[cfg(feature = "iterator")]
+    fn prefix_de_composite_key() {
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot/map.rs:522]: good test

**Review by ethanfrey — APPROVED** [review]
lgtm

File: packages/storage-plus/src/snapshot/map.rs (hunk 2 above)
Comment (maurolacy) [reply packages/storage-plus/src/snapshot/map.rs:249]: Yes, we can easily mess up here. A type system with something like composite types, whose parts or elements could be referred to independently would be nice, yes.

File: packages/storage-plus/src/snapshot/map.rs (hunk 1 above)
Comment (maurolacy) [reply packages/storage-plus/src/snapshot/map.rs:212]: Added a test / example anyway.

**Pull request merged**